### PR TITLE
Guard `depends_on :xcode` with `if OS.mac?`

### PR DESCRIPTION
**Pull request by jonchang**
- [x] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

https://github.com/Linuxbrew/brew/issues/796

### CODE REVIEW

File: Formula/dark-mode.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@ class DarkMode < Formula
   end
 
   depends_on :macos => :el_capitan
-  depends_on :xcode => :build
+  depends_on :xcode => :build if OS.mac?
   depends_on :macos
```
Comment (sjackman) [inline Formula/dark-mode.rb]: It's not necessary for this formula, which `depends_on :macos`

File: Formula/dshb.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ class Dshb < Formula
     sha256 "e0aa0f64ac02e9244fc59773a966d1ac5755dc4a4c91c0dbcd92633c4330f14b" => :el_capitan
   end
 
-  depends_on :xcode => ["8.0", :build]
+  depends_on :xcode => ["8.0", :build] if OS.mac?
   depends_on :macos
```
Comment (sjackman) [inline Formula/dshb.rb]: Ditto.

File: Formula/fmdiff.rb [hunk 3]
```diff
@@ -17,7 +17,7 @@ class Fmdiff < Formula
   # Needs FileMerge.app, which has been part of Xcode since Xcode 4 (OS X 10.7)
   # Prior to that it was included in the Developer Tools package.
   # "make" has logic for checking both possibilities.
-  depends_on :xcode if MacOS.version >= :lion
+  depends_on :xcode if MacOS.version >= :lion if OS.mac?
```
Comment (sjackman) [inline Formula/fmdiff.rb]: `brew audit` will complain about two `if` on the same line.
```ruby
depends_on :xcode if OS.mac? && MacOS.version >= :lion
```

File: Formula/freerdp.rb [hunk 4]
```diff
@@ -34,7 +34,7 @@ class Freerdp < Formula
     url "https://github.com/FreeRDP/FreeRDP/archive/4c69c3ea1489f09e1c3e698eaebd67b6d8d25785.tar.gz" # stable-1.1 branch as of Aug 13, 2016
     sha256 "1a4b03daca9bd82ef2e3827e1cdb469121f62aab1061f79ffa587babf9fe080c"
     version "1.1.0-beta1"
-    depends_on :xcode => :build # for "ibtool"
+    depends_on :xcode => :build # for "ibtool" if OS.mac?
```
Comment (sjackman) [inline Formula/freerdp.rb]: It's after a comment here.

File: Formula/graphviz.rb [hunk 5]
```diff
@@ -33,7 +33,7 @@ class Graphviz < Formula
   deprecated_option "with-pangocairo" => "with-pango"
 
   depends_on "pkg-config" => :build
-  depends_on :xcode => :build if build.with? "app"
+  depends_on :xcode => :build if build.with? "app" if OS.mac?
```
Comment (sjackman) [inline Formula/graphviz.rb]: Two ifs.

File: Formula/libmxml.rb [hunk 6]
```diff
@@ -13,7 +13,7 @@ class Libmxml < Formula
     sha256 "1cc925a2589ecf68f6b424294eb37f27086ce5c58698a2f9d9d642a139ecc708" => :x86_64_linux
   end
 
-  depends_on :xcode => :build # for docsetutil
+  depends_on :xcode => :build # for docsetutil if OS.mac?
```
Comment (sjackman) [inline Formula/libmxml.rb]: Comment

File: Formula/mplayershell.rb [hunk 7]
```diff
@@ -17,7 +17,7 @@ class Mplayershell < Formula
 
   depends_on "mplayer"
   depends_on :macos => :lion
-  depends_on :xcode => :build
+  depends_on :xcode => :build if OS.mac?
   depends_on :macos
```
Comment (sjackman) [inline Formula/mplayershell.rb]: macOS-only

File: Formula/pngpaste.rb [hunk 8]
```diff
@@ -14,7 +14,7 @@ class Pngpaste < Formula
   depends_on :macos => :el_capitan # needs NSBitmapImageFileTypePNG, etc.
 
   # Sierra's CLT is sufficient, but El Capitain's isn't
-  depends_on :xcode => [:build, "8.0"] if MacOS.version < :sierra
+  depends_on :xcode => [:build, "8.0"] if MacOS.version < :sierra and OS.mac?
```
Comment (sjackman) [inline Formula/pngpaste.rb]: `&&` rather than `and`
```ruby
depends_on :xcode => [:build, "8.0"] if OS.mac? && MacOS.version < :sierra
```

**Comment by jonchang** [comment]
Even if a formula has `depends_on :macos`, `brew info FORMULA` will still fail. E.g.

```
 $ brew info dark-mode
dark-mode: stable 2.0.1, HEAD
Control the macOS dark mode from the command-line
https://github.com/sindresorhus/dark-mode
Not installed
From: https://github.com/Linuxbrew/homebrew-core/blob/master/Formula/dark-mode.rb
==> Requirements
Error: undefined method `installed?' for OS::Mac::Xcode:Module
Did you mean?  instance_of?
```

File: Formula/sourcekitten.rb [hunk 9]
```diff
@@ -13,8 +13,8 @@ class Sourcekitten < Formula
   end
 
   depends_on :macos
-  depends_on :xcode => "6.0"
-  depends_on :xcode => ["9.0", :build]
+  depends_on :xcode => "6.0" if OS.mac?
+  depends_on :xcode => ["9.0", :build] if OS.mac?
```
Comment (sjackman) [inline Formula/sourcekitten.rb]: macOS-only

File: Formula/sourcery.rb [hunk 10]
```diff
@@ -10,8 +10,8 @@ class Sourcery < Formula
     sha256 "7119feb40c5beec48915c953e3e5af7d6614033d731647288d2e91c97f7e1264" => :high_sierra
   end
 
-  depends_on :xcode => "6.0"
-  depends_on :xcode => ["9.3", :build]
+  depends_on :xcode => "6.0" if OS.mac?
+  depends_on :xcode => ["9.3", :build] if OS.mac?
```
Comment (sjackman) [inline Formula/sourcery.rb]:
```ruby
if OS.mac?
…
end
```

File: Formula/swiftlint.rb [hunk 11]
```diff
@@ -15,8 +15,8 @@ class Swiftlint < Formula
   # Fixes the error: make: xcodebuild: Command not found
   depends_on :macos
 
-  depends_on :xcode => "8.0"
-  depends_on :xcode => ["9.0", :build]
+  depends_on :xcode => "8.0" if OS.mac?
+  depends_on :xcode => ["9.0", :build] if OS.mac?
```
Comment (sjackman) [inline Formula/swiftlint.rb]: Use an if block

File: Formula/xctool.rb [hunk 12]
```diff
@@ -12,7 +12,7 @@ class Xctool < Formula
     sha256 "cfd03ef1008a4a01b923e72fb7e79209d021d88ffe8bb89ac22307ba6af0cce0" => :el_capitan
   end
 
-  depends_on :xcode => "7.0"
+  depends_on :xcode => "7.0" if OS.mac?
   depends_on :macos
```
Comment (sjackman) [inline Formula/xctool.rb]: macOS-only

**Comment by sjackman** [comment]
Thanks, Joe! I've made a few comments.
What error do you see with Homebrew/brew if a formula has
```ruby
depends_on :xcode
```
?

**Comment by jonchang** [comment]
```
 $ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

 $ brew info xctool
xctool: stable 0.3.4, HEAD
Drop-in replacement for xcodebuild with a few extra features
https://github.com/facebook/xctool
Not installed
From: https://github.com/Linuxbrew/homebrew-core/blob/master/Formula/xctool.rb
==> Requirements
Error: undefined method `installed?' for OS::Mac::Xcode:Module
Did you mean?  instance_of?
Please report this bug:
  https://github.com/Linuxbrew/brew/wiki/troubleshooting
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirements/xcode_requirement.rb:14:in `xcode_installed_version'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirements/xcode_requirement.rb:6:in `block in <class:XcodeRequirement>'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:56:in `instance_eval'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:56:in `block in satisfied?'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:186:in `yielder'
/home/linuxbrew/.linuxbrew/Library/Homebrew/requirement.rb:56:in `satisfied?'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:198:in `block in decorate_requirements'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:196:in `collect'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:196:in `decorate_requirements'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:171:in `block in info_formula'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:168:in `map'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:168:in `info_formula'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:58:in `block in print_info'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:52:in `each'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:52:in `each_with_index'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:52:in `print_info'
/home/linuxbrew/.linuxbrew/Library/Homebrew/cmd/info.rb:41:in `info'
/home/linuxbrew/.linuxbrew/Library/Homebrew/brew.rb:87:in `<main>'

 $ git checkout guard_xcode
Switched to branch 'guard_xcode'
Your branch is up to date with 'jonchang/guard_xcode'.


 $ brew info xctool
xctool: stable 0.3.4, HEAD
Drop-in replacement for xcodebuild with a few extra features
https://github.com/facebook/xctool
Not installed
From: https://github.com/Linuxbrew/homebrew-core/blob/master/Formula/xctool.rb
==> Requirements
Required: macOS is required (uninstalled)
==> Options
--HEAD
	Install HEAD version
```

**Comment by jonchang** [comment]
I believe this is because `brew info` will look at all of the `depends_on`, and doing `depends_on :xcode` on non-macOS will :boom:

**Comment by jonchang** [comment]
In linuxbrew/brew this is addressed by just doing `return true` in `XcodeRequirement` https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/requirements/xcode_requirement.rb#L14

**Comment by sjackman** [comment]
Ah, got it

**Comment by sjackman** [comment]
Thanks for this work, Jon!

**Pull request merged**